Description: Add support for creating pid files.
Author: Maritn Pitt <mpitt@debian.org>
Bug: http://www.cups.org/str.php?L2465

Index: cups-1.6.0/scheduler/conf.c
===================================================================
--- cups-1.6.0.orig/scheduler/conf.c	2012-05-18 21:51:02.000000000 +0200
+++ cups-1.6.0/scheduler/conf.c	2012-07-27 16:55:12.250201963 +0200
@@ -175,7 +175,8 @@
 #endif /* HAVE_AUTHORIZATION_H */
   { "TempDir",			&TempDir,		CUPSD_VARTYPE_PATHNAME },
   { "Timeout",			&Timeout,		CUPSD_VARTYPE_TIME },
-  { "WebInterface",		&WebInterface,		CUPSD_VARTYPE_BOOLEAN }
+  { "WebInterface",		&WebInterface,		CUPSD_VARTYPE_BOOLEAN },
+  { "PidFile",			&PidFile,		CUPSD_VARTYPE_STRING }
 };
 #define NUM_VARS	(sizeof(variables) / sizeof(variables[0]))
 
@@ -586,6 +587,7 @@
   cupsdSetStringf(&ServerHeader, "CUPS/%d.%d", CUPS_VERSION_MAJOR,
                   CUPS_VERSION_MINOR);
   cupsdSetString(&StateDir, CUPS_STATEDIR);
+  cupsdSetString(&PidFile, "/var/run/cups/cupsd.pid");
 
   if (!strcmp(CUPS_DEFAULT_PRINTCAP, "/etc/printers.conf"))
     PrintcapFormat = PRINTCAP_SOLARIS;
Index: cups-1.6.0/scheduler/conf.h
===================================================================
--- cups-1.6.0.orig/scheduler/conf.h	2012-05-18 21:51:02.000000000 +0200
+++ cups-1.6.0/scheduler/conf.h	2012-07-27 16:53:22.590203425 +0200
@@ -242,6 +242,8 @@
 VAR int			SSLOptions		VALUE(CUPSD_SSL_NONE);
 					/* SSL/TLS options */
 #endif /* HAVE_SSL */
+VAR char		*PidFile		VALUE(NULL);
+					/* Debian CUPS pid file */
 
 #ifdef HAVE_LAUNCHD
 VAR int			LaunchdTimeout		VALUE(10);
Index: cups-1.6.0/scheduler/main.c
===================================================================
--- cups-1.6.0.orig/scheduler/main.c	2012-04-23 21:19:19.000000000 +0200
+++ cups-1.6.0/scheduler/main.c	2012-07-27 16:57:42.750199944 +0200
@@ -85,6 +85,8 @@
 static void		sigterm_handler(int sig);
 static long		select_timeout(int fds);
 static void		usage(int status) __attribute__((noreturn));
+int			write_pid(void);
+int			remove_pid(void);
 
 
 /*
@@ -587,6 +589,11 @@
   }
 #endif /* __sgi */
 
+  if (write_pid() == 0) {
+    cupsdLogMessage(CUPSD_LOG_ERROR, "Unable to write pid file");
+    return (1);
+  }
+
  /*
   * Initialize authentication certificates...
   */
@@ -1145,10 +1152,41 @@
 
   cupsdStopSelect();
 
+  remove_pid();
+
   return (!stop_scheduler);
 }
 
 
+/* 'write_pid()' - Write PID file.
+   'remove_pid()' - Delete PID file.
+*/
+int
+write_pid()
+{
+  FILE *f;
+  int fd;
+  int pid;
+  if (((fd = open(PidFile, O_RDWR|O_CREAT, 0644)) == -1)
+      || ((f = fdopen(fd, "r+")) == NULL) ) {
+    return 0;
+  }
+  pid = getpid();
+  if (!fprintf(f, "%d\n", pid)) {
+    close(fd);
+    return 0;
+  }
+  fflush(f);
+  close(fd);
+
+  return pid;
+}
+
+int
+remove_pid() {
+  return unlink(PidFile);
+}
+
 /*
  * 'cupsdAddString()' - Copy and add a string to an array.
  */
Index: cups-1.6.0/test/run-stp-tests.sh
===================================================================
--- cups-1.6.0.orig/test/run-stp-tests.sh	2012-05-21 17:35:25.000000000 +0200
+++ cups-1.6.0/test/run-stp-tests.sh	2012-07-27 16:53:22.590203425 +0200
@@ -400,6 +400,7 @@
 DocumentRoot $root/doc
 RequestRoot /tmp/cups-$user/spool
 TempDir /tmp/cups-$user/spool/temp
+PidFile /tmp/cups-$user/cupsd.pid
 MaxSubscriptions 3
 MaxLogSize 0
 AccessLog /tmp/cups-$user/log/access_log
