diff -urN cups.orig/filter/rastertopwg.c cups/filter/rastertopwg.c
--- cups.orig/filter/rastertopwg.c	2013-02-05 15:15:07.537013947 +0900
+++ cups/filter/rastertopwg.c	2013-02-05 15:17:46.589013928 +0900
@@ -26,6 +26,7 @@
 #include <cups/raster.h>
 #include <unistd.h>
 #include <fcntl.h>
+#include <signal.h>
 
 
 /*
@@ -61,6 +62,18 @@
   cups_option_t		*options = NULL;/* Options */
   const char		*val;		/* Option value */
 
+#if defined(HAVE_SIGACTION) && !defined(HAVE_SIGSET)
+    struct sigaction action;      /*   Actions for POSIX signals */
+#endif /*   HAVE_SIGACTION && !HAVE_SIGSET */
+    #ifdef HAVE_SIGSET
+    sigset(SIGPIPE, SIG_IGN);
+#elif defined(HAVE_SIGACTION)
+    memset(&action, 0, sizeof(action));
+    action.sa_handler = SIG_IGN;
+    sigaction(SIGPIPE, &action, NULL);
+#else
+    signal(SIGPIPE, SIG_IGN);
+#endif /*   HAVE_SIGSET */
 
   if (argc < 6 || argc > 7)
   {
Binary files cups.orig/.git/index and cups/.git/index differ
